posthog/api/signup.py

```diff
@@ -234,3 +207,156 @@ def get(self, request, *args, **kwargs):
+            )
+
+        request.session["organization_name"] = validated_data["organization_name"]
+        request.session["email_opt_in"] = validated_data["email_opt_in"]
+        request.session.set_expiry(3600)  # 1 hour to complete process
+        return {"continue_url": reverse("social:complete", args=[request.session["backend"]])}
+
+    def to_representation(self, instance: Any) -> Any:
+        return self.instance
+
+
+class SocialSignupViewset(generics.CreateAPIView):
+    serializer_class = SocialSignupSerializer
+    permission_classes = (CanCreateOrg,)
+
+
+class TeamInviteSurrogate:
+    """This reimplements parts of OrganizationInvite that enable compatibility with the old Team.signup_token."""
+
+    def __init__(self, signup_token: str):
+        team = Team.objects.select_related("organization").get(signup_token=signup_token)
+        self.organization = team.organization
+
+    def validate(*args, **kwargs) -> bool:
+        return True
+
+    def use(self, user: Any, *args, **kwargs) -> None:
+        user.join(organization=self.organization)
+
+
+class CompanyNameForm(forms.Form):
+    companyName = forms.CharField(max_length=64)
+    emailOptIn = forms.BooleanField(required=False)
+
+
+def finish_social_signup(request):
+    """
+    TODO: DEPRECATED in favor of posthog.api.signup.SocialSignupSerializer
+    """
+    if not get_can_create_org():
+        return redirect("/login?error=no_new_organizations")
+
+    if request.method == "POST":
+        form = CompanyNameForm(request.POST)
+        if form.is_valid():
+            request.session["organization_name"] = form.cleaned_data["companyName"]
+            request.session["email_opt_in"] = bool(form.cleaned_data["emailOptIn"])
+            return redirect(reverse("social:complete", args=[request.session["backend"]]))
+    else:
+        form = CompanyNameForm()
+    return render(request, "signup_to_organization_company.html", {"user_name": request.session["user_name"]})
+
+
+@partial
+def social_create_user(strategy: DjangoStrategy, details, backend, request, user=None, *args, **kwargs):
+    if user:
+        return {"is_new": False}
+    backend_processor = "social_create_user"
+    user_email = details["email"][0] if isinstance(details["email"], (list, tuple)) else details["email"]
+    user_name = details["fullname"] or details["username"]
+    strategy.session_set("user_name", user_name)
+    strategy.session_set("backend", backend.name)
+    from_invite = False
+    invite_id = strategy.session_get("invite_id")
+    if not invite_id:
+
+        # Check if the user is on a whitelisted domain
+        domain = user_email.split("@")[-1]
+        # TODO: Handle multiple organizations with the same whitelisted domain
+        domain_organization = Organization.objects.filter(domain_whitelist__contains=[domain]).first()
+
+        if domain_organization:
+            backend_processor = "domain_whitelist"
```

Comment:
"social_create_domain_whitelist" perhaps to tell the full story?

Reply:
I think I'll keep the name and add a comment on the code to clarify because as noted on Next Steps, one potential improvement could be supporting this behavior without third-party providers and just us validating their email directly.